File: foliage-rpc/src/main/java/io/foliage/netty/rpc/registry/ServiceRegistry.java
package io.foliage.netty.rpc.registry;

import lombok.extern.slf4j.Slf4j;
import org.apache.zookeeper.CreateMode;
import org.apache.zookeeper.KeeperException;
import org.apache.zookeeper.WatchedEvent;
import org.apache.zookeeper.Watcher;
import org.apache.zookeeper.ZooDefs;
import org.apache.zookeeper.ZooKeeper;
import org.apache.zookeeper.data.Stat;

import java.io.IOException;
import java.util.concurrent.CountDownLatch;

@Slf4j
public class ServiceRegistry {

    private final CountDownLatch latch = new CountDownLatch(1);
    private final String registryAddress;

    public ServiceRegistry(String registryAddress) {
        this.registryAddress = registryAddress;
    }

    public void register(String data) {
        if (data != null) {
            ZooKeeper zk = connectServer();
            if (zk != null) {
                AddRootNode(zk); // Add root node if not exist
                createNode(zk, data);
            }
        }
    }

    private ZooKeeper connectServer() {
        ZooKeeper zk = null;
        try {
            zk = new ZooKeeper(registryAddress, Constants.ZK_SESSION_TIMEOUT, new Watcher() {
                @Override
                public void process(WatchedEvent event) {
                    if (event.getState() == Event.KeeperState.SyncConnected) {
                        latch.countDown();
                    }
                }
            });
            latch.await();
        } catch (IOException | InterruptedException ex) {
            log.error("", ex);
        }
        return zk;
    }

    private void AddRootNode(ZooKeeper zk) {
        try {
            Stat s = zk.exists(Constants.ZK_REGISTRY_PATH, false);
            if (s == null) {
                zk.create(
                        Constants.ZK_REGISTRY_PATH,
                        new byte[0],
                        ZooDefs.Ids.OPEN_ACL_UNSAFE,
                        CreateMode.PERSISTENT
                );
            }
        } catch (KeeperException | InterruptedException ex) {
            log.error("", ex);
        }
    }

    private void createNode(ZooKeeper zk, String data) {
        try {
            byte[] bytes = data.getBytes();
            String path = zk.create(
                    Constants.ZK_DATA_PATH,
                    bytes,
                    ZooDefs.Ids.OPEN_ACL_UNSAFE,
                    CreateMode.EPHEMERAL_SEQUENTIAL
            );
            log.debug("create zookeeper node ({} => {})", path, data);
        } catch (KeeperException | InterruptedException ex) {
            log.error("", ex);
        }
    }
}

File: foliage-rpc/src/main/kotlin/org/lavenderx/foliage/nettyrpc/registry/ServiceRegistry.kt
package org.lavenderx.foliage.nettyrpc.registry

import org.apache.zookeeper.CreateMode
import org.apache.zookeeper.Watcher
import org.apache.zookeeper.ZooDefs
import org.apache.zookeeper.ZooKeeper
import org.lavenderx.foliage.nettyrpc.utils.loggerFor
import java.util.concurrent.CountDownLatch

class ServiceRegistry(private val registryAddress: String) {

    private val logger = loggerFor<ServiceRegistry>()
    private val latch = CountDownLatch(1)

    fun register(data: String?) {
        if (data != null) {
            val zk = connectServer()
            if (zk != null) {
                AddRootNode(zk) // Add root node if not exist
                createNode(zk, data)
            }
        }
    }

    private fun connectServer(): ZooKeeper? {
        var zk: ZooKeeper? = null
        try {
            zk = ZooKeeper(registryAddress, Constants.ZK_SESSION_TIMEOUT, Watcher { event ->
                if (event.state == org.apache.zookeeper.Watcher.Event.KeeperState.SyncConnected) {
                    latch.countDown()
                }
            })
            latch.await()
        } catch (e: Exception) {
            logger.error("", e)
        }

        return zk
    }

    private fun AddRootNode(zk: ZooKeeper) {
        try {
            val s = zk.exists(Constants.ZK_REGISTRY_PATH, false)
            if (s == null) {
                zk.create(
                        Constants.ZK_REGISTRY_PATH,
                        ByteArray(0),
                        ZooDefs.Ids.OPEN_ACL_UNSAFE,
                        CreateMode.PERSISTENT)
            }
        } catch (e: Exception) {
            logger.error("", e)
        }

    }

    private fun createNode(zk: ZooKeeper, data: String) {
        try {
            val bytes = data.toByteArray()
            val path = zk.create(
                    Constants.ZK_DATA_PATH,
                    bytes,
                    ZooDefs.Ids.OPEN_ACL_UNSAFE,
                    CreateMode.EPHEMERAL_SEQUENTIAL)
            logger.debug("create zookeeper node ({} => {})", path, data)
        } catch (e: Exception) {
            logger.error("", e)
        }

    }
}